# panel_requests

Abbreviations:
- CI: clinical indication
- PA: PanelApp (website at https://panelapp.genomicsengland.co.uk/)
- TD: NHS England National Genomic Test Directory (website at https://www.england.nhs.uk/publication/national-genomic-test-directories/)

Descriptions of the various gene and region metadata attributes can be found in the online PanelApp handbook at https://panelapp.genomicsengland.co.uk/media/files/PanelApp_Handbook_V18_120210506.pdf.

# Python
Please note this app requires Python version 3.10

# Setup
## Create or update database models
Make migrations if necessary and migrate to existing database.
```
python manage.py makemigrations requests_app
python manage.py migrate requests_app
```

## Populate the database
### 1. Insert data from PanelApp
The generic command for this is:
```
python manage.py seed panelapp all
```
- This command retrieves all signed-off panels from the PanelApp API, parses the data, and inserts it into the appropriate database models.
- It can be executed as-is and has no variable arguments.

### 2. Insert data from the National Genomic Test Directory
The generic command for this is:
```
python manage.py seed td <input_json>
```
- Example usage:
```
python manage.py seed td testing_files/230616_RD_TD_v5.json
```
- This command retrieves data from an output JSON file generated by "test directory parser", inserts the data into the appropriate database models, and link the clinical indication to the appropriate Panel record as specified
- The JSON file can be created from the original test directory MS Excel file using https://github.com/eastgenomics/test_directory_parser, and take the following format:

```json
{
  "td_source": "rare-and-inherited-disease-national-gnomic-test-directory-v5.1.xlsx",
  "config_source": "230401_RD",
  "date": "230616",
  "indications": [
    {
      "name": "Monogenic hearing loss",
      "code": "R67.1",
      "gemini_name": "R67.1_Monogenic hearing loss_P",
      "test_method": "WES or Large Panel",
      "panels": [
        "126"
      ],
      "original_targets": "Hearing loss (126)",
      "changes": "No change"
    },
  ]
}
```

The arguments for this command are:
- The name of the JSON file (which should be located within the same folder as the manage.py script)
  - An example of TD JSON file can be found under `testing_files/`


### 3. Seed transcript
```
python manage.py seed transcript --hgnc <path to hgnc.txt> --mane <path to mane.csv> --gff <path to parsed gff.tsv> --g2refseq <path to g2refseq.csv from HGMD database> --markname <path to markname.csv from HGMD database> <refgenome>
```
Arguments:
- `hgnc`: path to hgnc dump txt file
- `mane`: path to mane csv file
- `gff`: path to parsed gff.tsv (project-Fkb6Gkj433GVVvj73J7x8KbV:file-GF611Z8433Gk7gZ47gypK7ZZ)
- `g2refseq`: path to g2refseq table of HGMD database csv csv format
- `markname`: path to markanem table of HGMD database as csv format
- `refgenome`: reference genome e.g. 37 or GRCh37 or GRCh38

*HGMD database source can be found on DNAnexus (project-Fz4Q15Q42Z9YjYk110b3vGYQ:file-Fz4Q46842Z9z2Q6ZBjy7jVPY)
*g2refseq and markname are both exported table from HGMD database as csv format
*hgnc dump can be obtained from https://www.genenames.org/download/custom/

## Generate genepanel
```
Make a HGNC dump txt file here: https://www.genenames.org/download/custom/
Include columns:
- HGNC id
- Locus Type
- Approved Name

Without output pathway specified:
python manage.py genepanels --hgnc <hgnc dump path>
With output pathway specified:
python manage.py generate genepanels --hgnc testing_files/hgnc_dump_23052023.txt --output /home/jason/github/eris/testing_files
```

## Generate g2t	
```
python manage.py generate g2t --output <output pathway>
```


# Running unit tests

Unit tests are stored in the 'tests' directory, and can be run through 'manage.py':
```
python manage.py test tests
```
To run with line coverage:
```
pytest --cov=panel_requests panel_requests/requests_app/tests.py
```
Database-dependent tests use Django's unit testing library (based on unittest) to make, populate and tear down a temporary database based on models.py. Note that the test database will be named by prepending 'test_' to the value of the NAMEs in DATABASES.
